File: chap1_arrays_and_hashing/src/table.py
# table.py

"""HashTable with separate chaining for collision resolution."""

class Node:
    """One-line: Node storing a key-value pair in a linked list."""
    __slots__ = ('key', 'value', 'next')
    def __init__(self, key, value, nxt=None):
        self.key = key
        self.value = value
        self.next = nxt

class HashTable:
    """One-line: A simple hash table with chaining."""

    def __init__(self, capacity=16):
        """One-line: Initialize table with given capacity."""
        self.capacity = capacity
        self.size = 0
        self.buckets = [None] * capacity

    def _hash(self, key):
        """One-line: Compute bucket index for a key."""
        return hash(key) % self.capacity

    def set(self, key, value):
        """One-line: Insert or update key-value."""
        idx = self._hash(key)
        node = self.buckets[idx]
        while node:
            if node.key == key:
                node.value = value
                return
            node = node.next
        # insert new node at head
        self.buckets[idx] = Node(key, value, self.buckets[idx])
        self.size += 1

    def get(self, key):
        """One-line: Retrieve value for key or raise KeyError."""
        idx = self._hash(key)
        node = self.buckets[idx]
        while node:
            if node.key == key:
                return node.value
            node = node.next
        raise KeyError(key)

    def remove(self, key):
        """One-line: Remove key-value or raise KeyError."""
        idx = self._hash(key)
        prev = None
        node = self.buckets[idx]
        while node:
            if node.key == key:
                if prev:
                    prev.next = node.next
                else:
                    self.buckets[idx] = node.next
                self.size -= 1
                return
            prev, node = node, node.next
        raise KeyError(key)

File: chap1_arrays_and_hashing/src/table.c
// table.c

#include "table.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

// Simple hash: sum of characters modulo capacity
static unsigned int hash(const char *key, int capacity) {
    unsigned long hash = 0;
    for (; *key; key++) hash = hash * 31 + (unsigned char)*key;
    return hash % capacity;
}

// One-line: allocate and initialize hash table with buckets
HashTable *ht_create(int capacity) {
    HashTable *ht = malloc(sizeof(HashTable));
    ht->capacity = capacity;
    ht->size = 0;
    ht->buckets = calloc(capacity, sizeof(Entry *));
    return ht;
}

// One-line: free all entries and hash table
void ht_free(HashTable *ht) {
    for (int i = 0; i < ht->capacity; i++) {
        Entry *e = ht->buckets[i];
        while (e) {
            Entry *next = e->next;
            free(e->key);
            free(e);
            e = next;
        }
    }
    free(ht->buckets);
    free(ht);
}

// One-line: insert or update key-value
void ht_set(HashTable *ht, const char *key, int value) {
    unsigned int idx = hash(key, ht->capacity);
    Entry *e = ht->buckets[idx];
    while (e) {
        if (strcmp(e->key, key) == 0) {
            e->value = value; // update existing
            return;
        }
        e = e->next;
    }
    // not found: create a new entry and prepend
    Entry *new = malloc(sizeof(Entry));
    new->key = strdup(key);
    new->value = value;
    new->next = ht->buckets[idx];
    ht->buckets[idx] = new;
    ht->size++;
}

// One-line: retrieve value, set found flag
int ht_get(HashTable *ht, const char *key, int *found) {
    unsigned int idx = hash(key, ht->capacity);
    for (Entry *e = ht->buckets[idx]; e; e = e->next) {
        if (strcmp(e->key, key) == 0) {
            *found = 1;
            return e->value;
        }
    }
    *found = 0;
    return 0;
}

// One-line: remove key-value, free its memory
void ht_remove(HashTable *ht, const char *key) {
    unsigned int idx = hash(key, ht->capacity);
    Entry **ptr = &ht->buckets[idx];
    while (*ptr) {
        if (strcmp((*ptr)->key, key) == 0) {
            Entry *del = *ptr;
            *ptr = del->next;
            free(del->key);
            free(del);
            ht->size--;
            return;
        }
        ptr = &(*ptr)->next;
    }
}
